doc/tutorial/metaprogramming-arith.md

```diff
@@ -0,0 +1,290 @@
+## Arithmetic as an embedded domain-specific language
+
+Let's parse another classic grammar, the grammar of arithmetic expressions with
+addition, multiplication, integers, and variables.  In the process, we'll learn
+how to:
+
+- Convert identifiers such as `x` into strings within a macro.
+- Add the ability to "escape" the macro context from within the macro. This is useful to interpret identifiers with their _original_ meaning (predefined values)
+  instead of their new meaning within a macro (treat as a symbol).
+
+Let's begin with the simplest thing possible. We'll define an AST, and use operators `:+` and `:*` to denote
+building an arithmetic AST. We use `:+` and `:*` instead of `+` and `*` to avoid clashing names.
+
+
+Here's the AST that we will be parsing:
+
+```lean,ignore
+-- example on parsing arith language via macros
+inductive Arith: Type :=
+   | Add : Arith -> Arith -> Arith -- e :+ f
+   | Mul : Arith -> Arith -> Arith -- e :* f
+   | Int : Int -> Arith -- constant
+   | Symbol : String -> Arith -- variable
+```
+
+We declare a syntax category to describe the grammar that we will be parsing.
+See that we control the precedence of `:+` and `:*` by writing `syntax:50` for addition and `syntax:60` for multiplication,
+indicating that multiplication binds tighter than addition (higher the number, tighter the binding).
+This allows us to declare _precedence_ when defining new syntax.
+
+```lean,ignore
+declare_syntax_cat arith
+syntax term : arith -- int for Arith.Int
+syntax str : arith -- strings for Arith.Symbol
+syntax:50  arith ":+" arith : arith -- Arith.Add
+syntax:60 arith ":*" arith : arith -- Arith.Mul
+syntax "(" arith ")" : arith -- bracketed expressions
+```
+
+We define the macro `fromArith%` to translate the syntax category `arith` into an `Arith` inductive value that lives in `term`:
```

Comment:
Done, thanks for the stylistic note!

Reply:
I literally meant ``"`[Arith|" arith "]"`` :) . See, for example, https://github.com/IPDSnelting/tba-2021/blob/b6390e55b768423d3266969e81d19290129c5914/TBA/While/Com.lean#L58. I'm afraid the current `Arith|` notation is confusing because inside of quotations it looks like a "quotation kind" (e.g. `` `(level| ...) ``).

Reply:
Ah, I see :) Okay, I updated the notation to  `[Arith| ... ]`